### ISSUE TITLE
Conflict between processes sharing common stage directory

### ISSUE THREAD

**Issue by G-Ragghianti** [LLNL/spack#22386]
We recently started seeing what appears to be a conflict between spack processes of a user on the same host when they are building the same package.  This occurs in our CI system a lot.  The common stage directory will have the same directory name for a package's staged files so that when one process deletes the stage directory, it breaks the second process's build.  This occurs even when using separate spack installs. I'm going to be looking into a workaround by setting a custom stage directory, but I think the default should be to have a stage directory that is unique to the spack install. Thoughts?

Reproducing: Create two spack installs on the same host and run "spack install xz" on each at the same time.  You'll get various errors due to the conflict.

If it is agreed that a unique stage directory is desirable, I'm happy to create the patch.

Alternatively, the locking functionality could be slightly extended to work across separate spack installs (on the same system).

**Comment by haampie** [comment]
I think this is currently not possible, since we allow the user to do this: `spack stage spec && spack cd --stage-dir spec`. I think in some cases you might want to restage the sources too when you run with a debugger and need the sources again in their original location.

You can just use `spack -c config:build_stage:/tmp/$YOUR_UNIQUE_CI_JOB_ID stage zlib`

**Comment by G-Ragghianti** [comment]
To address your concerns, the stage directory could still be deterministically determined (e.g. by using a hash of $SPACK_HOME). I understand that I can change the build_stage, but I feel that the default should not include this conflict if it can be avoided.

I also just added a suggestion that extending the locking across spack installs would probably achieve the same result without the concerns that you mention above. This would probably be cleaner/easier.

**Comment by G-Ragghianti** [comment]
The stage process appears to use prefix_lock() (from database.py) in the stage directory for the package's spec, but this doesn't appear to be working as I would expect it to.

**Comment by G-Ragghianti** [comment]
@tgamblin @tldahlgren I see that you two are the most likely to know something about the locking system.  Do you have any thoughts on this (specifically about the prefix_lock() not properly locking the stage directories)? Thanks.

**Comment by tgamblin** [comment]
Are these two different spack instances with two different install trees?  Or two spack instances with the same install tree?

The latter should work, and if it doesn’t there is some issue with prefix lock. Prefix lock uses a lock file in which we lock per-package bits within the install tree.

If it’s the former, you’re right — that’s a good catch — we should set some stage directory unique to the spack installation. We can still determine that directory a priori (per @haampie’s use case above) but I’ll have to think about how to do it right. The stage directory is set in config so we will need some way to make it unique by spack instance within tmp. The workaround would be what you are doing (set the stage directory). 

Note that we made this last bit easier by adding `spack -c` recently for one off config options on the CLI.

**Comment by tgamblin** [comment]
I guess my other question is if it’s the former case above, why are you building the exact same package in two install trees? Just curious.

**Comment by G-Ragghianti** [comment]
This mostly happens on our CI system.  A lot of our builds share a spack instance, but some builds can't share or require a separate spack instance for testing.  

Edit: I just verified that the problem only occurs when building in two different spack trees on the same system.

**Comment by G-Ragghianti** [comment]
From my reading of how locking in the stage directory is intended to work, I think that it should be able to lock across separate spack install trees since it is using the lock file `/tmp/$USER/spack-stage/.lock`. If that is the case, then fixing the locking to work as intended would resolve all the issues (setting unique stage directories wouldn't be necessary).

Edit: Looking around line 311 of stage.py, it does not appears to use prefix_lock().

**Comment by tgamblin** [comment]
That would work, but I’m not sure that is the right solution, because it would slow down the build. In the shared install tree case, the two spack instances are ultimately trying to install the same thing. So, if one instance sees that the stage directory is locked, it can go do other work, and it will ultimately use the install that the other instance produced.

If there are two instances that share locks but not install trees, one will wait on the other, but it will ultimately have to build the package for itself as it can’t use the other’s installation.  So this ultimately slows down the build by making some work
impossible to do sooner, without doing other useful work.

**Comment by tgamblin** [comment]
> Edit: Looking around line 311 of stage.py, it does not appears to use prefix_lock().

Yes the prefix lock is used in `installer.py`, where the distributed locking algorithm is -- stages are created and destroyed while we hold the prefix lock

**Comment by G-Ragghianti** [comment]
> That would work, but I’m not sure that is the right solution, because it would slow down the build. In the shared install tree case, the two spack instances are ultimately trying to install the same thing. So, if one instance sees that the stage directory is locked, it can go do extra work, and it will ultimately use the install that the other instance produced.
> 
> If there are two instances that share locks but not install trees, one will wait on the other, but it will ultimately have to build the package for itself as it can’t use the other’s installation. So this ultimately slows down the build by making some work
> Impossible to do sooner, without doing other useful work.

That makes sense.  So we would need to hash something (full path) that is unique to the spack install tree in order to build a unique stage directory name, correct?

**Comment by haampie** [comment]
Actually let me reopen this, since #50697 doesn't deal with two spack instances & same stage dir.
